### PR TITLE
File name extractor draft

### PR DESCRIPTION
**Pull request by noahmacaskill**
Did a little work to finish off the file name extractor (which it seems you had pretty much finished up, just needed to tie it together). Tests are failing for a couple reasons:

- Sometimes the date isn't in the filename but is collected elsewhere, so the date in the fixture file doesn't match up with what is extracted (null in this case)
- There is sometimes a date discrepancy of a couple seconds between what is in the file name and in the header itself

Based on our earlier discussions, I think this may be the time to switch over to YAML? May have a go on this early next week, any thoughts/suggestions?

### CODE REVIEW

**Review by atruskie — APPROVED** [review]

**Comment by atruskie** [comment]
Good catch. Can you make sure the metadata.md file is updated too?

In terms of priorities; I'd like to see you transition into 

- making sure the metadata command just works on all our fixtures
- making sure the metadata command works for your entire set of test files (the set Doug and you set up 
- making sure the command works for Doug. Get him to run the command, review the output, and see whether it answers the questions he wants answered.

If amongst that you have time to do the YAML transcode of the fixtures data, then yeah go for it. As for notes: You'll probably find it easiest to do a straight conversion that is similarly flat, and then progressively add some nested structure.

**Comment by atruskie** [comment]
Oh I forgot to mention, the CI isn't passing. See if you can fix it please and then I'll merge.

File: test/Fixtures/Fixtures.yaml [hunk 1]
```diff
@@ -0,0 +1,902 @@
...
+        Speed: 50
+        Capacity: 124868608
+        WrCurrentVmin: 5
+        WrCurrentVmax: 5
+        WriteBlSize: 512
+        EraseBlSize: 65536
+      Sensor:
+        Firmware: 3.30
+        SerialNumber: 00008570
+        PowerSource: Ext-power
```
Comment (noahmacaskill) [inline test/Fixtures/Fixtures.yaml:902]: This is the structure I used for different cases in the YAML file. Conditionals is storing completely separate FixtureModel objects, to be used in specified cases (with specific extractors). Tests are now passing.

This wasn't the set up I was hoping to implement. I was hoping the conditionals could store only the values that needed to be changed when a different extractor is used, and that I could then apply those new values on to my existing FixtureModel, rather than creating a new one from scratch. Though I've tried different strategies, I haven't been able to get this to cooperate for a variety of reasons.

Figured I'd check in with what I have to see you think, or if you had any better ideas on how to configure the YAML for these cases.

**Comment by noahmacaskill** [comment]
On a note unrelated to this PR, I ran the extractor on all of our fixtures (all good there), then on some of Doug's file, where I ran into a weird little issue.

The code crashes [here](https://github.com/QutEcoacoustics/emu/blob/776d86a5031df0f653008dcc63e19a2cb68aae7b/src/MetadataUtility/Audio/Wave.cs#L300). It seems that for some reason the file size that is being written in the file header is slightly overshooting the actual file size, causing the program to crash on this check. When I comment that part out there are no issues, all looks great!

Thoughts on how to tackle this?

**Comment by atruskie** [comment]
> On a note unrelated to this PR, I ran the extractor on all of our fixtures (all good there), then on some of Doug's file, where I ran into a weird little issue.
> 
> The code crashes [here](https://github.com/QutEcoacoustics/emu/blob/776d86a5031df0f653008dcc63e19a2cb68aae7b/src/MetadataUtility/Audio/Wave.cs#L300). It seems that for some reason the file size that is being written in the file header is slightly overshooting the actual file size for some of our files, causing the program to crash. When I comment that part out there are no issues, all looks great!
> 
> Thoughts on how to tackle this?

let's talk about that issue tomorrow in our meeting. I have a suspicion on what the problem is. We'll inspect it together.

File: test/Fixtures/Fixtures.yaml [hunk 2]
```diff
@@ -0,0 +1,902 @@
...
+        UID: 009966
+        BuildDate: 2019-03-20
+        Gain: 50dB
+        Channel: 2
+        ChannelName: B
+        Sensitivity:
+    Voltage:
+    Gain:
+    Configuration:
+    BatteryLevel: 65p 3.79V
```
Comment (atruskie) [inline test/Fixtures/Fixtures.yaml:859]: Ok, already with YAML I'm seeing questionable values because it's so much easier to read in a diff.

I think we should split Battery percentage out from voltage. `BatteryLevel` should store an actual double in the range `[0.0, 1.0]` (the percentage as a unit value) and Battery voltage should be a separate field.

File this as a separate issue and address in another PR.

Comment (atruskie) [reply test/Fixtures/Fixtures.yaml:859]: There are similar issues with MemoryCard.ManufacturedDate and Microphone.Gain

File: test/Fixtures/Fixtures.yaml (hunk 1 above)
Comment (atruskie) [reply test/Fixtures/Fixtures.yaml:902]: I need to think about this more. 

I think we could probably merge conditionals into the `Process` field. Each process entry can be a name with an object. Empty object have no overrides, non-empty override specific fields. 

e.g.
[
```yaml
  Process:
    # override nothing
    - FilenameExtractor: {}
    # Override SD card serial number
    - FlacHeaderExtractor: 
        MemoryCards: 
            - SerialNumber: banana
```

I'm not sure on this though, so we'll discuss tomorrow.

**Review by atruskie — COMMENTED** [review]
I think conflating the YAML change in this PR is a bit of a mistake - two unrelated changes are now coupled.

For the file name stuff: good job. It's ready to merge.

For the fixture data/yaml stuff see if you can implement an override system that happens transparently to the tests.

For example: for every fixture record I'd like to be able N variants of a record where N is the number of extractors that can process a fixture. The variant's can be defined as overrides of gold standard/base record.

I also think it would be worthwhile to nest the record so fixture-related data is not messed in with audio recording related data. So the FixtureModel should be something like:

```yaml
- Name: 'hello'
  CanProcess: 
    - <name>
  Process:
   -  <name>:
       Vendor: Banana Labs
  Record:
    Vendor: ...
    StartDate: ...
    # ... 
```

**Comment by noahmacaskill** [comment]
In terms of the unrelated changes in the PR, getting the tests to pass was my main motivation. Since there are many unique cases when it comes to file name extraction, thought it made most sense to go YAML now rather than adding a set of conditional rows to the CSV, but I see what you mean. In hindsight, would have made more sense to convert to YAML before finishing up the filename extractor. Will discuss more in our meeting.

**Comment by noahmacaskill** [comment]
Doing the merger in YAML like we discussed didn't want to cooperate with YamlDotNet, basically since I'm deserializing into a FixtureModel, it was searching for '<<' as a property rather than realizing it is the merge key. Instead had to do the merger manually in C# [here](https://github.com/QutEcoacoustics/emu/blob/c650d67ead89fd93f5ffe3556750f5c8f91e02ed/test/MetadataUtility.Tests/TestHelpers/FixtureModel.cs#L93-L119).

I also removed the CanProcess column. With our YAML format, I don't see there being a case where something is in the CanProcess column but not in the Process column, since we will now just process one of our variations for any edge cases.

**Comment by atruskie** [comment]
> Instead had to do the merger manually in C# [here](https://github.com/QutEcoacoustics/emu/blob/c650d67ead89fd93f5ffe3556750f5c8f91e02ed/test/MetadataUtility.Tests/TestHelpers/FixtureModel.cs#L93-L119).

We didn't talk about this in the meeting but I really wanted to avoid that. There a lot of edge cases involved with doing this stuff in C# because of it's strongly typed nature.

As such if you go that route I'd strongly encourage you to find a library that does the overwrites for you. Maybe AutoMapper? https://automapper.org/
 

As for the YAML dot net side of things the solution is to pass the input through a merging parser before passing it along to the deserializer. 

Example: https://github.com/QutEcoacoustics/audio-analysis/blob/85ac7985bcddbd5bec0bbb9b83057106d1442b17/src/Acoustics.Shared/Yaml.cs#L52-L59

File: test/MetadataUtility.Tests/TestHelpers/FixtureModel.cs [hunk 3]
```diff
@@ -115,6 +90,34 @@ public TargetInformation ToTargetInformation(IFileSystem fileSystem)
...
         }
 
+        /// <summary>
+        /// Overwrite certain values in the recording.
+        /// Used for edge cases with some extractor.
+        /// </summary>
+        /// <param name="overwriteRecording">The FixtureModel containing the values to overwrite.</param>
+        public void ApplyOverwrites(FixtureModel overwriteRecording)
+        {
+            foreach (string overwrite in overwriteRecording.OverwriteProperties)
```
Comment (atruskie) [inline test/MetadataUtility.Tests/TestHelpers/FixtureModel.cs:100]: Ok I just worked out why you explicitly listed properties together.

If you can't the yaml stream merging working (as per https://github.com/QutEcoacoustics/audio-analysis/blob/85ac7985bcddbd5bec0bbb9b83057106d1442b17/src/Acoustics.Shared/Yaml.cs#L52-L59)

Then I'd suggest you deserialize `Process ` as a Dictionary<string, object>`. You can then use graph traversal and type testing to add properties to the record.

**Review by atruskie — CHANGES_REQUESTED** [review]
We're deep in the murky waters of C# and strong typing here. Reflection based mechanism are almost always the wrong choice for a variety of reasons.

I'd like to see you get MergingParser solution working.  Otherwise we'll let this rest and I'll clean it up at some point in the future. I'm not happy with the solution but I don't want you getting bogged down in this either.

**Comment by noahmacaskill** [comment]
Merging parser worked, yeah this is much cleaner.
